src/quadiron-c.cpp

```diff
@@ -0,0 +1,450 @@
+    std::vector<int> wanted_data_vec(fec->n_data, 0);
+    std::vector<int> wanted_idxs_vec(fec->n_outputs, 0);
+    int metadata_size = quadiron_fnt32_get_metadata_size(fecp, block_size);
+    bool res;
+    int ret;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret =
+                    parities_props[i].from_buffer(metadata, metadata_size / 4);
+                if (-1 == ret)
+                    return -1;
+            }
+            parities_vec[i] = parity[i] + metadata_size;
+        }
+    } else {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (!missing_idxs[i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(data[i]);
+                int ret =
+                    parities_props[i].from_buffer(metadata, metadata_size / 4);
+                if (-1 == ret)
+                    return -1;
+            }
+            parities_vec[i] = data[i] + metadata_size;
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[fec->n_data + i].from_buffer(
+                    metadata, metadata_size / 4);
+                if (-1 == ret)
+                    return -1;
+            }
+            parities_vec[fec->n_data + i] = parity[i] + metadata_size;
+        }
+    }
+
+    int need_decode = 0;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        /*
+         * Easy case where the target is a data then simply decode
+         */
+        if (destination_idx < fec->n_data) {
+            wanted_idxs_vec[destination_idx] = 1;
+            res = fec->decode_blocks_vertical(
+                data_vec,
+                parities_vec,
+                parities_props,
+                missing_idxs_vec,
+                wanted_idxs_vec,
+                block_size);
+            if (!res) {
+                ret = -1;
+                goto end;
+            }
+            ret = 0;
+            goto end;
+        }
+    }
+
+    /*
+     * At this point we want a parity to be reconstructed
+     * If systematic we may need to decode if a data is missing.
+     * If non-systematic we always need to decode
+     */
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (missing_idxs[i]) {
+                need_decode = 1;
+                wanted_data_vec[i] = 1;
+                data_vec[i] = static_cast<uint8_t*>(malloc(block_size));
```

Comment:
could use `new`, since we are in C++ code.

Or, since it seems that this `data_vec` is only used internally, use a std::vector => then you don't need your goto dance to be sure that memory is freed at the end (yeah BTW if anything raise an exception here you'll leak: goto won't save you => in C++, where virtually/pessimistically every line can disrupt the control flow, the only way to prevent leak is to use RAII, goto only works for pure C code).

Reply:
Can you suggest a code change ? because I don't see how it can auto-free ?

Reply:
By using a std::vector, as seen with you.

Reply:
done